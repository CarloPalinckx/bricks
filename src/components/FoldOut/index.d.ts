import { Component } from 'react';
declare type ContentProps = {
    contentHeight?: StateType['contentHeight'];
    isOpen: PropsType['isOpen'];
};
declare type StateType = {
    contentHeight?: number;
};
declare type PropsType = {
    isOpen: boolean;
};
declare class FoldOut extends Component<PropsType, StateType> {
    private contentRef;
    private resizeObserver;
    constructor(props: PropsType);
    componentDidMount(): void;
    componentWillUnmount(): void;
    render(): JSX.Element;
}
export default FoldOut;
export { ContentProps, StateType, PropsType };
