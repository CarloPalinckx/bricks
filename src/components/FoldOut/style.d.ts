import { default as _R } from 'react';
import { StyledComponentClass as _S } from 'styled-components';
import { ContentProps } from '.';
declare const StyledFoldOut: _S<_R.ClassAttributes<HTMLDivElement> & _R.HTMLAttributes<HTMLDivElement> & ContentProps, any, _R.ClassAttributes<HTMLDivElement> & _R.HTMLAttributes<HTMLDivElement> & ContentProps>;
export default StyledFoldOut;
