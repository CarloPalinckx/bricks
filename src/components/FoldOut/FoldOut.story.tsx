import { storiesOf } from '@storybook/react';
import React, { Component } from 'react';
import FoldOut from '.';
import Button from '../Button';
import Spacer from '../Spacer';
import Text from '../Text';

type StateType = {
    isOpen:boolean;
};

const demoContent = `
    Now this is the story all about how
    My life got flipped, turned upside down
    And I'd like to take a minute just sit right there
    I'll tell you how I became the prince of a town called Bel-air

    In west Philadelphia born and raised
    On the playground where I spent most of my days
    Chilling out, maxing, relaxing all cool
    And all shooting some b-ball outside of the school
    When a couple of guys, they were up to no good
    Started making trouble in my neighbourhood
    I got in one little fight and my mom got scared
    And said "You're moving with your auntie and uncle in Bel-air"

    I whistled for a cab and when it came near the
    License plate said "fresh" and had a dice in the mirror
    If anything I could say that this cab was rare
    But I thought nah, forget it, yo homes to Bel-air!

    I pulled up to a house about seven or eight
    And I yelled to the cabby "Yo, homes smell you later!"
    Looked at my kingdom I was finally there
    To sit on my throne as the prince of Bel-air
`;

class DemoComponent extends Component<{}, StateType> {
    public constructor(props:{}) {
        super(props);

        this.state = {
            isOpen: false,
        };
    }

    public render():JSX.Element {
        return (
            <div>
                <FoldOut isOpen={this.state.isOpen}>
                    <Spacer bottom={12} offsetType="inner">
                        <Text>{demoContent}</Text>
                    </Spacer>
                </FoldOut>
                <Button
                    action={():void => this.setState({
                        isOpen: !this.state.isOpen
                    })}
                    title="open FoldOut"
                    variant="secondary"
                >
                    Toggle
                </Button>
            </div>
        );
    }
}

storiesOf('FoldOut', module)
    .add('With a toggle', () => (
        <DemoComponent />
    ));
