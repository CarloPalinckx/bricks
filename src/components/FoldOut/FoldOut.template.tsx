import React, { Component } from 'react';
import StyledFoldOut from './FoldOut.style';

type ContentProps = {
    contentHeight?:StateType['contentHeight'];
    isOpen:PropsType['isOpen'];
};

type StateType = {
    contentHeight?:number;
};

type PropsType = {
    isOpen:boolean;
};

class FoldOut extends Component<PropsType, StateType> {
    private contentRef:HTMLDivElement | null;

    public constructor(props:PropsType) {
        super(props);

        this.state = {
            contentHeight: undefined,
        };
    }

    private updateHeight = ():void => {
        if (this.contentRef !== null) {
            this.setState({
                contentHeight: this.contentRef.offsetHeight,
            });
        }
    }

    public componentDidMount():void {
        window.addEventListener('resize', this.updateHeight);
        this.updateHeight();
    }

    public componentWillUnmount():void {
        window.removeEventListener('resize', this.updateHeight);
    }

    public render():JSX.Element {
        return (
            <StyledFoldOut isOpen={this.props.isOpen} contentHeight={this.state.contentHeight}>
                <div ref={(ref):void => { this.contentRef = ref; }}>
                    {this.props.children}
                </div>
            </StyledFoldOut>
        );
    }
}

export default FoldOut;
export {
    ContentProps,
    StateType,
    PropsType,
};
