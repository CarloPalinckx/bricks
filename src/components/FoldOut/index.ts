export { default } from './FoldOut.template';
