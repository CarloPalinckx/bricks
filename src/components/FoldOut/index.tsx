import React, { Component } from 'react';
import ResizeObserver from 'resize-observer-polyfill';
import StyledFoldOut from './style';

type ContentProps = {
    contentHeight?: StateType['contentHeight'];
    isOpen: PropsType['isOpen'];
};

type StateType = {
    contentHeight?: number;
};

type PropsType = {
    isOpen: boolean;
};

class FoldOut extends Component<PropsType, StateType> {
    private contentRef: HTMLDivElement;
    private resizeObserver: ResizeObserver;

    public constructor(props: PropsType) {
        super(props);

        this.state = {
            contentHeight: undefined,
        };

        try {
            this.resizeObserver = new ResizeObserver((entries: Array<ResizeObserverEntry>): void => {
                entries.forEach((entry: ResizeObserverEntry): void => {
                    this.setState({
                        contentHeight: entry.contentRect.height,
                    });
                });
            });
        } catch (error) {
            console.warn(`
                ResizeObserver is not available in this environment.
                 Folding animation will be unavailable. Fallback used.
            `);
        }
    }

    public componentDidMount(): void {
        try {
            this.resizeObserver.observe(this.contentRef);
        } catch {
            // no-op when ResizeObserver is not available.
        }
    }

    public componentWillUnmount(): void {
        try {
            this.resizeObserver.unobserve(this.contentRef);
        } catch {
            // no-op when ResizeObserver is not available.
        }
    }

    public render(): JSX.Element {
        return (
            <StyledFoldOut isOpen={this.props.isOpen} contentHeight={this.state.contentHeight}>
                <div
                    ref={(ref): void => {
                        this.contentRef = ref as HTMLDivElement;
                    }}
                >
                    {this.props.children}
                </div>
            </StyledFoldOut>
        );
    }
}

export default FoldOut;
export { ContentProps, StateType, PropsType };
