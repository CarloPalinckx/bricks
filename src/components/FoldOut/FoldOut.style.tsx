import { default as _R } from 'react';
import styled, { StyledComponentClass as _S } from 'styled-components';
import { ContentProps } from './FoldOut.template';

const StyledFoldOut = styled.div`
    transition: height 300ms cubic-bezier(0.5, 0, 0.1, 1);
    overflow: hidden;
    height: ${(props: ContentProps): string => {
        if (!props.isOpen) {
            return '0';
        }

        if (props.contentHeight !== undefined) {
            return `${props.contentHeight}px`;
        }

        return 'auto';
    }};
`;

export default StyledFoldOut;
