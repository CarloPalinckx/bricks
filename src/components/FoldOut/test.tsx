import React from 'react';
import renderer from 'react-test-renderer';
import ResizeObserver from 'resize-observer-polyfill';
import FoldOut from '../FoldOut';

/* tslint:disable */
(console.warn as any).mockImplementation(() => {});
/* tslint:enable */

const renderOptions = {
    createNodeMock: (): Object => ({
        contentRect: {
            height: 900,
        },
    }),
};

describe('FoldOut', () => {
    it('should have a height when open', () => {
        const foldOut = renderer.create(<FoldOut isOpen />, renderOptions).toJSON();

        expect(foldOut).toMatchSnapshot();
    });

    it('should have no height when closed', () => {
        const foldOut = renderer.create(<FoldOut isOpen={false} />, renderOptions).toJSON();

        expect(foldOut).toMatchSnapshot();
    });

    it('should use the fallback and warn once ResizeObserver is not available or throws', () => {
        (ResizeObserver as jest.Mock<ResizeObserver>).mockImplementationOnce(() => {
            throw TypeError;
        });

        const foldOut = renderer.create(<FoldOut isOpen />, renderOptions).toJSON();

        expect(foldOut).toMatchSnapshot();
        /* tslint:disable */
        expect(console.warn).toHaveBeenCalled();
        /* tslint:enable */
    });
});
