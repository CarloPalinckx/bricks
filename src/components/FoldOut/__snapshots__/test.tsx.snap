// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoldOut should have a height when open 1`] = `
<div
  className="dmhAZb"
>
  <div />
</div>
`;

exports[`FoldOut should have no height when closed 1`] = `
<div
  className="bSsVjP"
>
  <div />
</div>
`;

exports[`FoldOut should use the fallback and warn once ResizeObserver is not available or throws 1`] = `
<div
  className="kZzNJL"
>
  <div />
</div>
`;
